### PR TITLE
Add parent pointers, better DOM nesting warning

### LINKED ISSUE #2849: Save component back references in dev mode
For the purposes of providing dev-tools support and providing back traces in error messages/warnings, it might be useful to keep back references to a component's parent.  This trace information will cut down the amount of time required to find the component(s) responsible for throwing warnings.

### PR DESCRIPTION
**Pull request by spicyj**
Fixes #2849, and I think we need something like these parent pointers in order to make our new algorithm work.

### CODE REVIEW

**Comment by spicyj** [comment]
@sebmarkbage I assume this is all gross and you'll tell me the proper way to do this. :)

File: src/browser/ui/__tests__/ReactDOMComponent-test.js [hunk 1]
```diff
@@ -697,9 +702,8 @@ describe('ReactDOMComponent', function() {
       expect(console.error.calls.length).toBe(1);
       expect(console.error.calls[0].args[0]).toBe(
         'Warning: validateDOMNesting(...): <tr> cannot appear as a child of ' +
-        '<table> in this context (div > table). Add a <tbody> to your code ' +
-        'to match the DOM tree generated by the browser. Check the render ' +
-        'method of `Foo`.'
+        '<table> in this context (div > Foo > table > Row > tr). Add a ' +
```
Comment (spicyj) [inline src/browser/ui/__tests__/ReactDOMComponent-test.js]: This message is so much better though.

**Comment by syranide** [comment]
Interesting, apart from some remnant bug or two, this should allow us to revisit the idea of dropping `data-reactid` without resorting to hacks I think...

**Comment by sebmarkbage** [comment]
I was looking at adding back references from composites to their DOM container so that we can reconcile updates from that point. We might need back references for every node in the tree for other purposes though. We should figure out the optimal number of back references since there seems to be a number of variations with various tradeoffs.

The most powerful one is simply a reference to the next immediate parent, then you have to walk it to find a DOM parent.

**Comment by spicyj** [comment]
Yes, this uses a reference to the next immediate parent.

**Comment by sebmarkbage** [comment]
Wouldn't the owner tree from both parent and child to the common owner ancestor be enough and provide more concise context?

`... > Foo > table > ... > Row > tr`

**Comment by spicyj** [comment]
Maybe so. It's not always the parent though; sometimes a higher ancestor makes it invalid so we need to store away each of the ancestor DOM components that could restrict the valid children so that we can then go back and crawl up the owners.

**Comment by sebmarkbage** [comment]
How about storing something like `{ 'p': internalInstance, 'table': internalInstance }` which tells the children of the context that they're currently inside a `p` and `table` tag and also gives access to owner trees?

**Comment by spicyj** [comment]
Yeah, okay. Done in #3719.

**Comment by spicyj** [comment]
(Abandoning this for now in favor of #3719.)

**Pull request closed without merging**